### ISSUE TITLE
`failed to list repos with archived label` error when syncing a Bitbucket Data Center code host

### ISSUE THREAD

**Issue by mike-r-mclaughlin** [sourcegraph/sourcegraph#40308]
- **Sourcegraph version:** 3.42.0
- **Reported by:**  https://github.com/sourcegraph/accounts/issues/3

#### Details

Code host sync fails with:

`error":"failed to list repos with archived label: Get \https://bitbucket.net/rest/api/1.0/labels/archived/labeled?REPOSITORY=&limit=1000\: net/http: request canceled (Client.Timeout exceeded while awaiting headers)"`

Full logs:

```
{"SeverityText":"ERROR","Timestamp":1660213485409315516,"InstrumentationScope":"service.ObservedSource","Caller":"repos/observability.go:99","Function":"[github.com/sourcegraph/sourcegraph/internal/repos](http://github.com/sourcegraph/sourcegraph/internal/repos).(*observedSource).ListRepos.func1","Body":"source.list-repos","Resource":{"service.name":"repo-updater","service.version":"3.42.0","[service.instance.id](http://service.instance.id/)":"repo-updater-IIId-xcte5-2rr2s"},"Attributes":{"error":"failed to list repos with archived label: Get \https://bitbucket.net/rest/api/1.0/labels/archived/labeled?REPOSITORY=&limit=1000\: net/http: request canceled (Client.Timeout exceeded while awaiting headers)"}}
{"SeverityText":"ERROR","Timestamp":1660213485409709442,"InstrumentationScope":"service.syncer","Caller":"repos/syncer.go:600","Function":"[github.com/sourcegraph/sourcegraph/internal/repos](http://github.com/sourcegraph/sourcegraph/internal/repos).(*Syncer).SyncExternalService","Body":"error from codehost","Resource":{"service.name":"repo-updater","service.version":"3.42.0","[service.instance.id](http://service.instance.id/)":"repo-updater-IIId-xcte5-2rr2s"},"Attributes":{"svc":{"name":"III Cluster09 - Bitbucket Server","id":3},"seen":0,"error":"failed to list repos with archived label: Get \https://bitbucket.net/rest/api/1.0/labels/archived/labeled?REPOSITORY=&limit=1000\: net/http: request canceled (Client.Timeout exceeded while awaiting headers)"}}
```

Bitbucket server versions:
* 5.16.3
* 7.21.0

This seems to block syncing for the code host.

/cc @jplahn @ryphil

**Comment by mike-r-mclaughlin** [comment]
From @stefanhengl : 
> The newer version showed the error. However, we didn't confirm yet on our side that 5.16.3 doesn't show this error.

**Comment by indradhanush** [comment]
## Update 2022-08-15

**TL;DR**: I've found a reliable way to reproduce this and have some thoughts on what should be the next step.

### Steps to reproduce

- Port forward to the `bitbucket` service on the replica instance:

```
kubectl -n ns-sourcegraph port-forward svc/bitbucket 7991:http
``` 

- Add a code host config from your local `dev` instance to this instance by adding the following config to `$PROJECT_ROOT/../dev-private/enterprise/dev/external-services-config.json`


```
  "BITBUCKETSERVER": [
    {
      "url": "http://localhost:7991",
      "token": "<redacted>",
      "username": "admin",
      "projectKeys": [
        "A1",
        "A5",
        "A8",
        "AA",
        "AB",
        "AC",
        "AD",
        "AE",
        "AF",
        "AG"
      ]
    }
  ]
```


- Trigger a manual sync and you should be able to see the error reported in this issue


- In case the issue is not see, follow the steps in https://github.com/sourcegraph/sourcegraph/issues/40153#issuecomment-1214672976 to reproduce that issue first and then you should see this error as well from your local dev instance



### Thoughts on further investigation

I couldn't get around to trying this yet, so this would have been my next steps:

1. The customer saw the issue reliably in `7.21.0` and our replica instance is also running `7.21.3` which is where we are also able to reproduce this issue. But the customer did not see this issue on their bitbucket server running `5.16.3`. Is this a correlation? We should run bitbucket server with `5.16.3` and try the above steps and see if we can reproduce this or not.


2. Additionally, this feature was implemented quite a few years back in response to a customer's requirement based on how they were simulating GitHub's archived repo functionality

- In the customer that is facing this issue and the replica instance that we have created we do not have any repo with this label at the moment

- Maybe we can add this label to one of the repos and try this network request again

- Does it continue to fail?


### Suggested temporary fix

- At the moment we fail completely when [the code to fetch repos with the archived label](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@ffa89deaf50ee9f14f1bf15def8d44c04122f218/-/blob/internal/repos/bitbucketserver.go?L199-203) returns an error. Instead, we should log this as a warning and continue the sync and think about a way of communicating a "partial" sync to the user.

**Comment by ryanslade** [comment]
> At the moment we fail completely when [the code to fetch repos with the archived label](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@ffa89deaf50ee9f14f1bf15def8d44c04122f218/-/blob/internal/repos/bitbucketserver.go?L199-203) returns an error. Instead, we should log this as a warning and continue the sync and think about a way of communicating a "partial" sync to the user.

Unfortunately I don't think this temporary fix would work since we actually return a map with the ids of all archived repos, then lower down we use this map to mark repos as archived if they're in the map. If we ignore the error we'll sync all repos without tagging them as archived which is probably not what we want.

**Comment by indradhanush** [comment]
> Just to double check, can we ask their admin to manually try and get the list of archived repos and check how long that takes?

@ryanslade My bad that I forgot to add that we already tried this over a call last week with the customer. 

We tried visiting the URL in the browser as well as making the curl request from within the `repo-updater` pod of the customer's production cluster and in both cases we saw the same response as you mentioned above - the one with the `NoSuchLabelException`.

**Comment by ryanslade** [comment]
Can we check if the customer has rate limiting enabled on their bitbucket instance? I'm wondering if maybe the newer versions of Bitbucket have rate limiting enabled which is causing it to block?

**Comment by mike-r-mclaughlin** [comment]
@ryanslade they said they likely have some rate limiting enabled but they are not sure and the team that runs these servers is not very responsive so we likely won't know for sure for a few weeks.  

They also told me they don't think Bitbucket has a concept of archived repositories.  Is this label something new?  

When we saw this on the call, it wasn't clear if it was blocking repositories on this code host from syncing or not.  Do you know if this is blocking repos on this code host from syncing?

**Comment by ryanslade** [comment]
Correct, bitbucket doesn't have the concept of archived repos so some customers use labels to get around this. 

This may be a red herring and could just be a symptom of general network timeouts. The label request is one of the first ones we make to BitBucket so would be the first to appear in any logs if there are issues communicating with the BitBucket instance.

**Comment by ryanslade** [comment]
It would stop the current current sync as we treat it as a fatal error.

**Comment by mike-r-mclaughlin** [comment]
ok.  it seems to happen regularly.  anything else I can gather from the logs to help us?

**Comment by indradhanush** [comment]
I don't have anything else that I can think of at the moment. I'm working on setting up a `5.16.3` Bitbucket server to rule out if this issue is limited to `7.x`. 

Post that I will also verify with `8.x` - the latest available Bitbucket server version.

**Comment by mike-r-mclaughlin** [comment]
Customer has not seen this error since removing the one problematic Bitbucket server from Sourcegraph.  When we re-added that server back to Sourcegraph, we did not see the exception.   However, we did see the error from #40449 and (presumably as a result of this) no repositories were added from the problematic Bitbucket instance.   So, I'm not sure if we didn't see the `label not found` exception because of #40449 or if it's not happening any more.

**Comment by indradhanush** [comment]
@mike-r-mclaughlin With more investigation we're more certain that this error is coming from the network layer, the timeout indicating that maybe there is an unusual network setup that doesn't allow too many requests at once. 

Also, our default rate limits are set to 8 requests per second (28800 per hour). Maybe there's some rule on their end that is limiting this once a threshold (presumably their internal threshold) is breached and the connections don't get through to the Bitbucket server. 

If they start seeing this error again, I would suggest them to configure a lower rate limit, maybe half of the default - 14400 per hour (4 requests per second) in the code host configuration.

I'd be curious to see if the error goes away then.

**Comment by mike-r-mclaughlin** [comment]
@indradhanush it happened one more time on a new server:
> This error is generated for the new cluster we were trying to add. Once we get this error it looks like the sync does not happen.
> Logs
```
{"SeverityText":"WARN","Timestamp":1662990515185132314,"InstrumentationScope":"sync_worker.worker.Handle","Caller":"workerutil/worker.go:409","Function":"[github.com/sourcegraph/sourcegraph/internal/workerutil](http://github.com/sourcegraph/sourcegraph/internal/workerutil).(*Worker).handle","Body":"Marked record as errored","Resource":{"service.name":"repo-updater","service.version":"3.43.2","[service.instance.id](http://service.instance.id/)":"repo-updater-67fffcc544-2bpbp"},"Attributes":{"error":"fetching from code host Bitbucket Cluster XX Server: failed to list repos with archived label: Get \https://bibucket.com/rest/api/1.0/labels/archived/labeled?REPOSITORY=&limit=1000\: net/http: request canceled (Client.Timeout exceeded while awaiting headers)"}}
{"SeverityText":"ERROR","Timestamp":1662990515185227911,"InstrumentationScope":"sync_worker.worker.Handle","Caller":"workerutil/worker.go:382","Function":"[github.com/sourcegraph/sourcegraph/internal/workerutil](http://github.com/sourcegraph/sourcegraph/internal/workerutil).(*Worker).handle.func1","Body":"operation.error","Resource":{"service.name":"repo-updater","service.version":"3.43.2","[service.instance.id](http://service.instance.id/)":"repo-updater-67fffcc544-2bpbp"},"Attributes":{"count":1,"elapsed":300.026788688,"error":"fetching from code host Bitbucket Cluster XX Server: failed to list repos with archived label: Get \https://bibucket.com/rest/api/1.0/labels/archived/labeled?REPOSITORY=&limit=1000\: net/http: request canceled (Client.Timeout exceeded while awaiting headers)"}}
```

**Comment by indradhanush** [comment]
@mike-r-mclaughlin IIRC, this was resolved once the customer opened up port 443 for their SG instance right? Can we close this?
